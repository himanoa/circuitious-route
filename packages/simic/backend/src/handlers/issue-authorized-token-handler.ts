import { SQL, SQLStatement } from "sql-template-strings";
import * as Express from "express";
import { UserNotFoundError, LoginIdNotFoundError } from "../error";

type Token = string;

export const issueAuthorizedTokenHandler: (deps: {
  executeQuery: Promise<(sql: string | SQLStatement) => Promise<any>>;
  sign: (obj: { discordId: string }) => Token;
  generateRandomString: () => string;
}) => (req: Express.Request, res: Express.Response) => Promise<void> = (
  deps
) => async (req, res) => {
  const executeQuery = await deps.executeQuery;
  try {
    if (req.params && req.params.loginId === undefined) {
      throw new LoginIdNotFoundError("Login id is not found");
    }
    const [user] = await executeQuery(
      SQL`SELECT * FROM users WHERE current_login_id = ${req.params.loginId};`
    );
    if (!user) {
      throw new UserNotFoundError(`User is not found`);
    }
    const refreshToken = await executeQuery(
      SQL`SELECT * FROM refresh_tokens WHERE discord_id = ${user.discord_id} AND activated = 1;`
    );

    if (refreshToken.length === 0) {
      const refreshToken = deps.generateRandomString();
      await executeQuery(
        SQL`INSERT INTO refresh_tokens (discord_id, token, activated, created_at) VALUES (${user.discord_id}, ${refreshToken}, 1, strftime('%s', 'now'));`
      );
      res.status(201).json({
        refreshToken: refreshToken,
        accessToken: deps.sign({ discordId: user.discord_id }),
      });
    } else {
      res.status(401).json({});
    }
  } catch (err) {
    if (err.name === "LoginIdNotFoundError") {
      res.status(400).json({});
    } else if (err.name === "UserNotFoundError") {
      res.status(401).json({ error: err });
    } else {
      res.status(500).json({
        error: err,
      });
    }
  }
};
