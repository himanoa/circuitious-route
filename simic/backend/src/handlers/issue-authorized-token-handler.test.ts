import { issueAuthorizedTokenHandler } from "./issue-authorized-token-handler"
import { SQLStatement } from "sql-template-strings"
import * as Response from "jest-express/lib/response"
import * as Request from "jest-express/lib/request"

describe("issueAuthorizedTokenHandler", () => {
  describe("when exist discordId", () => {
    describe("when exist refreshToken", () => {
      const executedQuery: string[] = []
      const dummyDeps = {
        executeQuery: Promise.resolve((sql: SQLStatement | string) => {
          if(typeof sql === "string") {
            executedQuery.push(sql)
            return Promise.resolve([1])
          }
          executedQuery.push(sql.sql)
          return Promise.resolve([1])
        }),
        generateRandomString: () => "foobar",
        sign: () => "signed strings"
      }
      
      it("should be valid", async () => {
        const request = new Request.Request("https://example.com")
        request.setBody({
          discordId: 123123123
        })
        const response = new Response.Response()
        await issueAuthorizedTokenHandler(dummyDeps)(request as any, response as any)
        expect(executedQuery).toMatchSnapshot()
        expect(response.status).toBeCalledWith(401)
        expect(response.body).toMatchSnapshot()
      })
    })

    describe("when not exist refreshToken", () => {
      const executedQuery: string[] = []
      const dummyDeps = {
        executeQuery: Promise.resolve((sql: SQLStatement | string) => {
          if(typeof sql === "string") {
            executedQuery.push(sql)
            return Promise.resolve([1])
          }
          executedQuery.push(sql.sql)
          if(/^SELECT \* FROM refresh_tokens/.test(sql.sql)) {
            return Promise.resolve([])
          }
          return Promise.resolve([1])
        }),
        generateRandomString: () => "foobar",
        sign: () => "signed strings"
      }
      
      it("should be valid", async () => {
        const request = new Request.Request("https://example.com")
        request.setBody({
          discordId: 123123123
        })
        const response = new Response.Response()
        await issueAuthorizedTokenHandler(dummyDeps)(request as any, response as any)
        expect(executedQuery).toMatchSnapshot()
        expect(response.status).toBeCalledWith(201)
        expect(response.body).toMatchSnapshot()
      })
    })
    })
  })
  describe("when not exist discordId", () => {
    const executedQuery: string[] = []
    const dummyDeps = {
      executeQuery: Promise.resolve((sql: SQLStatement | string) => {
        if(typeof sql === "string") {
          executedQuery.push(sql)
          return Promise.resolve([])
        }
        executedQuery.push(sql.sql)
        return Promise.resolve([])
      }),
      generateRandomString: () => "foobar",
      sign: () => "signed strings"
    }
    
    it("should be valid", async () => {
      const request = new Request.Request("https://example.com")
      request.setBody({
        discordId: 123123123
      })
      const response = new Response.Response()
      await issueAuthorizedTokenHandler(dummyDeps)(request as any, response as any)
      expect(executedQuery).toMatchSnapshot()
      expect(response.status).toBeCalledWith(201)
      expect(response.body).toMatchSnapshot() 
    })
  })

  describe("when invalid body", () => {
    const executedQuery: string[] = []
    const dummyDeps = {
      executeQuery: Promise.resolve((sql: SQLStatement | string) => {
        if(typeof sql === "string") {
          executedQuery.push(sql)
          return Promise.resolve([])
        }
        executedQuery.push(sql.sql)
        return Promise.resolve([])
      }),
      generateRandomString: () => "foobar",
      sign: () => "signed strings"
    }
    
    it("should be error", async () => {
      const request = new Request.Request("https://example.com")
      request.setBody({
        invalid: 123123123
      })
      const response = new Response.Response()
      await issueAuthorizedTokenHandler(dummyDeps)(request as any, response as any)
      expect(executedQuery).toMatchSnapshot()
      expect(response.status).toBeCalledWith(400)
      expect(response.body).toMatchSnapshot() 
    })
  })

  describe("when db error", () => {
    const executedQuery: string[] = []
    const dummyDeps = {
      executeQuery: Promise.resolve(() => {
        throw Error("DB")
      }),
      generateRandomString: () => "foobar",
      sign: () => "signed strings"
    }
    
    it("should be error", async () => {
      const request = new Request.Request("https://example.com")
      request.setBody({
        discordId: 123123123
      })
      const response = new Response.Response()
      await issueAuthorizedTokenHandler(dummyDeps)(request as any, response as any)
      expect(executedQuery).toMatchSnapshot()
      expect(response.status).toBeCalledWith(500)
      expect(response.body).toMatchSnapshot() 
    })
  })
