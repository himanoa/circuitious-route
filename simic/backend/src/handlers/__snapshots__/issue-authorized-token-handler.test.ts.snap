// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`issueAuthorizedTokenHandler when exist discordId when exist refreshToken should be valid 1`] = `
Array [
  "SELECT * FROM users WHERE current_login_id = ?;",
  "SELECT * FROM refresh_tokens WHERE discord_id = ? AND activated = 1;",
]
`;

exports[`issueAuthorizedTokenHandler when exist discordId when exist refreshToken should be valid 2`] = `Object {}`;

exports[`issueAuthorizedTokenHandler when exist discordId when not exist refreshToken should be valid 1`] = `
Array [
  "SELECT * FROM users WHERE current_login_id = ?;",
  "SELECT * FROM refresh_tokens WHERE discord_id = ? AND activated = 1;",
  "INSERT INTO refresh_tokens (discord_id, token, activated, created_at) VALUES (?, ?, 1, strftime('%s', 'now'));",
]
`;

exports[`issueAuthorizedTokenHandler when exist discordId when not exist refreshToken should be valid 2`] = `
Object {
  "accessToken": "signed strings",
  "refreshToken": "foobar",
}
`;

exports[`when db error should be error 1`] = `Array []`;

exports[`when db error should be error 2`] = `
Object {
  "error": [Error: DB],
}
`;

exports[`when invalid params should be error 1`] = `Array []`;

exports[`when invalid params should be error 2`] = `Object {}`;
